File: TREES/maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    int dfs(TreeNode* root,int curMax,int curMin)
    {
        if(root == NULL)
        {
            return curMax-curMin;
        }

        curMax = max(root->val,curMax);
        curMin = min(root->val,curMin);

        int left = dfs(root->left,curMax,curMin);
        int right = dfs(root->right,curMax,curMin);

        return max(left,right);

    }

    int maxAncestorDiff(TreeNode* root) {
        
        if(!root)
        {
            return 0;
        }
        return dfs(root,root->val,root->val);
    }
};

File: Searching_Algorithms/maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        def get_max(root,cur_max,cur_min):
            if not(root):
                return abs(cur_max-cur_min)
            cur_max = max(cur_max,root.val)
            cur_min  = min(root.val,cur_min)
            left_max = get_max(root.left,cur_max,cur_min)
            right_max = get_max(root.right,cur_max,cur_min)
            
            return max(left_max,right_max)
        return get_max(root,root.val,root.val)
